[WIP] Add a new op that composes all of the dynamic shape logic

navahgar opened the pull request:
Differential Revision: D32320082

Krovatkin commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -1299,9 +1324,97 @@ void FuseTensorExprs(
+  // };
+
+  return [=](Stack& stack) {
+    static std::unordered_map<std::string, std::shared_ptr<tensorexpr::TensorExprKernel>> cached_kernels;
```
is `cached_kernels` accessed from multiple threads?

navahgar replied:
Yes. In the SR workflow, that can potentially happen. I will have to update this to be thread-safe.

Krovatkin commented on torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:
```diff
@@ -376,5 +389,31 @@ RegisterOperators reg_guard({
+  //   interpreter.run(stack);
+  //   return 0;
+  // };
+  return [=](Stack& stack) {
```
nit: maybe be explicit of things we are copying

eellison replied:
wouldnt this be instantiated a code object/Interpreter state on every invocation of the op? i think we want to create an InterpreterState outside of the lambda

navahgar replied:
If I create the `InterpreterState` outside the lambda, then the context, specifically `frames` in `InterpreterStateImpl` is garbled when the lambda is getting executed. That is the reason I had to move the instantiation to inside lambda. 

If there is a way to instantiate that outside and reuse it for every invocation of the op, that would be awesome. And would definitely help perf

Krovatkin replied:
@navahgar `InterpreterState` is pretty lightweight or used to be. On the other hand if you can cache `Code` object construction, that could be pretty significant savings since `Code` translates JIT graphs into interpreter instructions, does register allocation and a bunch of other passes. Both, legacy and profiling executors, cache `Code` objects in `ExecutionPlan` for this exact reason.

You could do 

```cpp
Code code(graph, ""); // code should be copyable.
return [code](Stack& stack) {
```

`InterpreterState` should be doing mainly two things: 
* push the Code object on the `frames` stack
* Reserve enough spaces for virtual registers

Both should be fairly low-overhead. Not sure it would be worthwhile to reuse the `InterpreterState` vs `Code` tbh

cc: @eellison

eellison replied:
This is supposed to be an optimized kernel execution... so even small overheads matter here. Would be great to find a way to avoid the duplicate instantiation

Krovatkin commented on torch/csrc/jit/runtime/jit_trace.cpp:
```diff
@@ -290,6 +290,14 @@ std::shared_ptr<Graph> TraceGraph(std::shared_ptr<Graph> graph, Stack& stack) {
     ni->setType(ni->type());
     td.old_to_new_[inp] = ni;
   }
+  // type inputs
```
nit: you already checked in a fix ?

navahgar replied:
That diff (D32382958) just landed. Will update this PR.

navahgar committed:
[StaticRuntime] Refactor StaticModule to pass in sample inputs.

Summary:
This diff refactors StaticModule and its uses to pass in sample
inputs.

Test Plan: buck run mode/opt //caffe2/caffe2/fb/predictor:pytorch_predictor_test

Differential Revision: D32320084

fbshipit-source-id: 91fd44319f0fc3cffe28038dbd687fd2fc49dafe

navahgar committed:
[StaticRuntime] Add TensorExpr fusion with dynamic shapes in SR

Summary:
This diff adds TensorExpr fusion with dynamic shapes in SR. This
includes tracing the input graph with sample inputs, and then performing fusion
with generalization to get fused graphs with dynamic shapes.

Test Plan: buck run mode/opt //caffe2/caffe2/fb/predictor:pytorch_predictor_test

Differential Revision: D32320088

fbshipit-source-id: a9a925a7d0e93151ccd61ed6e19f07bb4e2bbf66

navahgar committed:
[jit] Add support for dynamic shape fusion in JIT.

Summary:
This diff adds support for dynamic shape fusion in JIT. This is done
by performing fusion with the static shapes observed on the first run,
generalizing the fused subgraphs and generating code for the generalized fused
subgraphs with dynamic shapes.

Test Plan: buck test mode/dev-nosan //caffe2/test/cpp/jit:jit

Differential Revision: D32781307

fbshipit-source-id: ffbc0e1afc708a0cd7916a27519530d37d99927c

navahgar committed:
[jit] Add a new op that composes all of the dynamic shape logic (#68207)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/68207

This diff adds a new op, `TensorExprDynamicGroup`, that composes all the logic behind running a dynamic shaped fused node. This includes a guard instruction that checks for conditions, a conditional that calls the fused node or the fallback graph depending on the guard.

Test Plan:
```
buck test mode/dev-nosan //caffe2/test/cpp/jit:jit
```

Differential Revision: D32320082

fbshipit-source-id: d6d25d39acaf7be2873af140ab53da4ba4153d06
